// @ngx-gallery/core/esm5/ngx-gallery-core.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { Gallery, GalleryRef, GalleryComponent, GalleryAction, ImageSize, LoadingStrategy, ThumbnailsPosition, ThumbnailsMode, SlidingDirection, GalleryItemType, ImageItem, VideoItem, IframeItem, YoutubeItem, GalleryModule } from './public_api';
export { GalleryCoreComponent as ɵe } from './lib/components/gallery-core.component';
export { GalleryCounterComponent as ɵg } from './lib/components/gallery-counter.component';
export { GalleryDotsComponent as ɵd } from './lib/components/gallery-dots.component';
export { GalleryItemComponent as ɵj } from './lib/components/gallery-item.component';
export { GalleryNavComponent as ɵc } from './lib/components/gallery-nav.component';
export { GallerySliderComponent as ɵf } from './lib/components/gallery-slider.component';
export { GalleryThumbComponent as ɵi } from './lib/components/gallery-thumb.component';
export { GalleryThumbsComponent as ɵh } from './lib/components/gallery-thumbs.component';
export { GalleryIframeComponent as ɵm } from './lib/components/templates/gallery-iframe.component';
export { GalleryImageComponent as ɵk } from './lib/components/templates/gallery-image.component';
export { GalleryVideoComponent as ɵl } from './lib/components/templates/gallery-video.component';
export { LazyDirective as ɵn } from './lib/directives/lazy.directive';
export { TapClickDirective as ɵo } from './lib/directives/tap-click.directive';
export { GALLERY_CONFIG as ɵa } from './lib/utils/gallery.token';
