// @ngx-gallery/core/esm2015/lib/components/templates/gallery-image.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, Output, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
export class GalleryImageComponent {
    /**
     * @param {?} _sanitizer
     */
    constructor(_sanitizer) {
        this._sanitizer = _sanitizer;
        /**
         * Stream that emits when an error occurs
         */
        this.error = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get imageLoaded() {
        return !!this.loadedImage;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.loadingIcon) {
            this.loadingSvg = this._sanitizer.bypassSecurityTrustHtml(this.loadingIcon);
        }
    }
}
GalleryImageComponent.decorators = [
    { type: Component, args: [{
                selector: 'gallery-image',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `
    <ng-container [lazyImage]="src" (loaded)="loadedImage = $event" (error)="error.emit($event)">
      <div *ngIf="loadedImage; else loading"
           class="g-image-item"
           [style.backgroundImage]="loadedImage"></div>
    </ng-container>

    <ng-template #loading>
      <i *ngIf="loadingSvg; else thumbLoading" class="g-loading" [innerHTML]="loadingSvg"></i>
      <ng-template #thumbLoading>
        <div class="g-thumb-loading"></div>
      </ng-template>
    </ng-template>
  `
            },] },
];
/** @nocollapse */
GalleryImageComponent.ctorParameters = () => [
    { type: DomSanitizer }
];
GalleryImageComponent.propDecorators = {
    src: [{ type: Input }],
    loadingIcon: [{ type: Input }],
    error: [{ type: Output }],
    imageLoaded: [{ type: HostBinding, args: ['class.g-image-loaded',] }]
};
if (false) {
    /** @type {?} */
    GalleryImageComponent.prototype.loadedImage;
    /** @type {?} */
    GalleryImageComponent.prototype.loadingSvg;
    /** @type {?} */
    GalleryImageComponent.prototype.src;
    /** @type {?} */
    GalleryImageComponent.prototype.loadingIcon;
    /**
     * Stream that emits when an error occurs
     * @type {?}
     */
    GalleryImageComponent.prototype.error;
    /** @type {?} */
    GalleryImageComponent.prototype._sanitizer;
}
