// @ngx-gallery/lightbox/esm5/lib/lightbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostBinding, HostListener, Optional, ChangeDetectionStrategy } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { Location } from '@angular/common';
import { Subscription } from 'rxjs';
import { lightboxAnimations } from './lightbox.animation';
var LightboxComponent = /** @class */ (function () {
    function LightboxComponent(sanitizer, location) {
        var _this = this;
        this.sanitizer = sanitizer;
        /**
         * Subscription to changes in the user's location.
         */
        this._locationChange$ = Subscription.EMPTY;
        // Close the Lightbox when the location changes
        if (location) {
            this._locationChange$ = location.subscribe(function () { return _this.overlayRef.detach(); });
        }
    }
    /** Dispose the overlay when exit animation is done */
    /**
     * Dispose the overlay when exit animation is done
     * @param {?} e
     * @return {?}
     */
    LightboxComponent.prototype.onMouseLeave = /**
     * Dispose the overlay when exit animation is done
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.toState === 'void') {
            this.overlayRef.dispose();
        }
    };
    /**
     * @return {?}
     */
    LightboxComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._locationChange$.unsubscribe();
    };
    LightboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lightbox',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [lightboxAnimations.slideLightbox],
                    styles: ["::ng-deep lightbox{position:relative;display:block;width:1100px;height:800px;max-width:94vw;max-height:90vh}::ng-deep lightbox gallery{overflow:hidden;margin:0;display:block;width:100%;height:100%;box-shadow:0 11px 15px -7px rgba(0,0,0,.2),0 24px 38px 3px rgba(0,0,0,.14),0 9px 46px 8px rgba(0,0,0,.12)}::ng-deep .g-backdrop{background-color:rgba(0,0,0,.9)}::ng-deep .fullscreen{width:100%}::ng-deep .fullscreen ::ng-deep lightbox{max-width:unset;max-height:unset;position:fixed;top:0;left:0;bottom:0;right:0;height:100%;width:100%}::ng-deep .g-overlay{margin:auto}@media only screen and (max-width:480px){::ng-deep .g-overlay{width:100%}::ng-deep .g-overlay ::ng-deep lightbox{max-width:unset;max-height:unset;position:fixed;top:0;left:0;bottom:0;right:0;height:100%;width:100%}}::ng-deep .cdk-global-overlay-wrapper,::ng-deep .cdk-overlay-container{pointer-events:none;top:0;left:0;height:100%;width:100%}::ng-deep .cdk-overlay-container{position:fixed;z-index:1000}::ng-deep .cdk-global-overlay-wrapper{display:flex;position:absolute;z-index:1000}::ng-deep .cdk-overlay-pane{position:absolute;pointer-events:auto;box-sizing:border-box;z-index:1000}::ng-deep .cdk-overlay-backdrop{position:absolute;top:0;bottom:0;left:0;right:0;z-index:1000;pointer-events:auto;-webkit-tap-highlight-color:transparent;transition:opacity .4s cubic-bezier(.25,.8,.25,1);opacity:0}::ng-deep .cdk-overlay-backdrop.cdk-overlay-backdrop-showing{opacity:.48}::ng-deep .cdk-overlay-dark-backdrop{background:rgba(0,0,0,.6)}::ng-deep .cdk-overlay-transparent-backdrop{background:0 0}::ng-deep .cdk-global-scrollblock{position:fixed;width:100%;overflow-y:scroll}::ng-deep .g-btn-close{position:absolute;right:.9em;top:.9em;z-index:60;cursor:pointer;width:20px;height:20px}@media only screen and (max-width:480px){::ng-deep .g-btn-close{right:.7em;top:.7em}}"],
                    template: "\n    <gallery [id]=\"id\" [destroyRef]=\"false\" [skipInitConfig]=\"true\">\n      <i class=\"g-btn-close\" aria-label=\"Close\" (click)=\"overlayRef.detach()\"\n         [innerHTML]=\"sanitizer.bypassSecurityTrustHtml(closeIcon)\"></i>\n    </gallery>\n  "
                },] },
    ];
    /** @nocollapse */
    LightboxComponent.ctorParameters = function () { return [
        { type: DomSanitizer },
        { type: Location, decorators: [{ type: Optional }] }
    ]; };
    LightboxComponent.propDecorators = {
        slideAnimation: [{ type: HostBinding, args: ['@slideLightbox',] }],
        onMouseLeave: [{ type: HostListener, args: ['@slideLightbox.done', ['$event'],] }]
    };
    return LightboxComponent;
}());
export { LightboxComponent };
if (false) {
    /**
     * Gallery ref id
     * @type {?}
     */
    LightboxComponent.prototype.id;
    /**
     * Overlay ref to close the lightbox
     * @type {?}
     */
    LightboxComponent.prototype.overlayRef;
    /**
     * close button svg data
     * @type {?}
     */
    LightboxComponent.prototype.closeIcon;
    /**
     * Subscription to changes in the user's location.
     * @type {?}
     */
    LightboxComponent.prototype._locationChange$;
    /**
     * Use slide animation on opening and closing the lightbox
     * @type {?}
     */
    LightboxComponent.prototype.slideAnimation;
    /** @type {?} */
    LightboxComponent.prototype.sanitizer;
}
