// @ngx-gallery/core/esm2015/lib/directives/lazy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { Subject, zip, fromEvent } from 'rxjs';
import { tap, switchMap } from 'rxjs/operators';
export class LazyDirective {
    constructor() {
        this._worker$ = new Subject();
        this.loaded = new EventEmitter();
        this.error = new EventEmitter();
        /** @type {?} */
        const img = new Image();
        this._worker$.pipe(switchMap((imageSrc) => {
            // Image is loading
            this.loaded.emit(null);
            // Stop previously loading
            img.src = imageSrc;
            /** @type {?} */
            const loadSuccess = fromEvent(img, 'load').pipe(tap(() => {
                this.loaded.emit(`url(${imageSrc})`);
            }));
            /** @type {?} */
            const loadError = fromEvent(img, 'error').pipe(tap(() => {
                this.error.emit(new Error(`[lazyImage]: The image ${imageSrc} did not load`));
                this.loaded.emit(null);
            }));
            return zip(loadSuccess, loadError);
        })).subscribe();
    }
    /**
     * @param {?} imagePath
     * @return {?}
     */
    set lazyImage(imagePath) {
        this.loadImage(imagePath);
    }
    /**
     * @param {?} imagePath
     * @return {?}
     */
    loadImage(imagePath) {
        this._worker$.next(imagePath);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._worker$.complete();
    }
}
LazyDirective.decorators = [
    { type: Directive, args: [{
                selector: '[lazyImage]'
            },] },
];
/** @nocollapse */
LazyDirective.ctorParameters = () => [];
LazyDirective.propDecorators = {
    lazyImage: [{ type: Input, args: ['lazyImage',] }],
    loaded: [{ type: Output }],
    error: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LazyDirective.prototype._worker$;
    /** @type {?} */
    LazyDirective.prototype.loaded;
    /** @type {?} */
    LazyDirective.prototype.error;
}
