// @ngx-gallery/core/esm5/lib/directives/lazy.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { Subject, zip, fromEvent } from 'rxjs';
import { tap, switchMap } from 'rxjs/operators';
var LazyDirective = /** @class */ (function () {
    function LazyDirective() {
        var _this = this;
        this._worker$ = new Subject();
        this.loaded = new EventEmitter();
        this.error = new EventEmitter();
        /** @type {?} */
        var img = new Image();
        this._worker$.pipe(switchMap(function (imageSrc) {
            // Image is loading
            // Image is loading
            _this.loaded.emit(null);
            // Stop previously loading
            img.src = imageSrc;
            /** @type {?} */
            var loadSuccess = fromEvent(img, 'load').pipe(tap(function () {
                _this.loaded.emit("url(" + imageSrc + ")");
            }));
            /** @type {?} */
            var loadError = fromEvent(img, 'error').pipe(tap(function () {
                _this.error.emit(new Error("[lazyImage]: The image " + imageSrc + " did not load"));
                _this.loaded.emit(null);
            }));
            return zip(loadSuccess, loadError);
        })).subscribe();
    }
    Object.defineProperty(LazyDirective.prototype, "lazyImage", {
        set: /**
         * @param {?} imagePath
         * @return {?}
         */
        function (imagePath) {
            this.loadImage(imagePath);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} imagePath
     * @return {?}
     */
    LazyDirective.prototype.loadImage = /**
     * @param {?} imagePath
     * @return {?}
     */
    function (imagePath) {
        this._worker$.next(imagePath);
    };
    /**
     * @return {?}
     */
    LazyDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._worker$.complete();
    };
    LazyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[lazyImage]'
                },] },
    ];
    /** @nocollapse */
    LazyDirective.ctorParameters = function () { return []; };
    LazyDirective.propDecorators = {
        lazyImage: [{ type: Input, args: ['lazyImage',] }],
        loaded: [{ type: Output }],
        error: [{ type: Output }]
    };
    return LazyDirective;
}());
export { LazyDirective };
if (false) {
    /** @type {?} */
    LazyDirective.prototype._worker$;
    /** @type {?} */
    LazyDirective.prototype.loaded;
    /** @type {?} */
    LazyDirective.prototype.error;
}
