// @ngx-gallery/core/esm5/lib/directives/tap-click.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
var TapClickDirective = /** @class */ (function () {
    function TapClickDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.tapClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TapClickDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setTapEvent();
    };
    /** Enable gestures if hammer is loaded */
    /**
     * Enable gestures if hammer is loaded
     * @return {?}
     */
    TapClickDirective.prototype.setTapEvent = /**
     * Enable gestures if hammer is loaded
     * @return {?}
     */
    function () {
        var _this = this;
        if (typeof Hammer !== 'undefined') {
            // Use hammer.js tap event
            this.mc = new Hammer(this.el.nativeElement);
            this.mc.on('tap', function () {
                if (!_this.tapClickDisabled) {
                    _this.tapClick.emit(null);
                }
            });
        }
        else {
            // Use normal click event
            this.clickListener = this.renderer.listen(this.el.nativeElement, 'click', function (e) {
                e.stopPropagation();
                if (!_this.tapClickDisabled) {
                    _this.tapClick.emit(null);
                }
            });
        }
    };
    /**
     * @return {?}
     */
    TapClickDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.mc) {
            this.mc.destroy();
        }
        else {
            this.clickListener();
        }
    };
    TapClickDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[tapClick]'
                },] },
    ];
    /** @nocollapse */
    TapClickDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    TapClickDirective.propDecorators = {
        tapClickDisabled: [{ type: Input }],
        tapClick: [{ type: Output }]
    };
    return TapClickDirective;
}());
export { TapClickDirective };
if (false) {
    /** @type {?} */
    TapClickDirective.prototype.mc;
    /** @type {?} */
    TapClickDirective.prototype.clickListener;
    /** @type {?} */
    TapClickDirective.prototype.tapClickDisabled;
    /** @type {?} */
    TapClickDirective.prototype.tapClick;
    /** @type {?} */
    TapClickDirective.prototype.el;
    /** @type {?} */
    TapClickDirective.prototype.renderer;
}
