// @ngx-gallery/gallerize/bundles/ngx-gallery-gallerize.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@ngx-gallery/core'), require('@ngx-gallery/lightbox'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@ngx-gallery/gallerize', ['exports', '@angular/core', '@angular/common', '@ngx-gallery/core', '@ngx-gallery/lightbox', 'rxjs', 'rxjs/operators'], factory) :
    (factory((global['ngx-gallery'] = global['ngx-gallery'] || {}, global['ngx-gallery'].gallerize = {}),global.ng.core,global.ng.common,null,null,global.rxjs,global.rxjs.operators));
}(this, (function (exports,core,common,core$1,lightbox,rxjs,operators) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var GallerizeDirective = /** @class */ (function () {
        function GallerizeDirective(_el, _gallery, _lightbox, _renderer, platform, _galleryCmp) {
            this._el = _el;
            this._gallery = _gallery;
            this._lightbox = _lightbox;
            this._renderer = _renderer;
            this._galleryCmp = _galleryCmp;
            /**
             * Default gallery id
             */
            this._galleryId = 'lightbox';
            /**
             * The selector used to query images elements
             */
            this.selector = 'img';
            // Set gallerize mode
            if (common.isPlatformBrowser(platform)) {
                this._mode = _galleryCmp ? "gallery" /* Gallery */ : "detector" /* Detector */;
            }
        }
        /**
         * @return {?}
         */
        GallerizeDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this._galleryId = this.gallerize || this._galleryId;
                /** @type {?} */
                var ref = this._gallery.ref(this._galleryId);
                switch (this._mode) {
                    case "detector" /* Detector */:
                        this.detectorMode(ref);
                        break;
                    case "gallery" /* Gallery */:
                        this.galleryMode(ref);
                }
            };
        /**
         * @return {?}
         */
        GallerizeDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                switch (this._mode) {
                    case "detector" /* Detector */:
                        this._detector$.complete();
                        this._observer$.disconnect();
                        break;
                    case "gallery" /* Gallery */:
                        this._itemClick$.unsubscribe();
                        this._itemChange$.unsubscribe();
                }
            };
        /**
         * Adds a click event to each gallery items to make it opens in in lightbox
         * @param {?} galleryRef
         * @return {?}
         */
        GallerizeDirective.prototype.galleryMode = /**
         * Adds a click event to each gallery items to make it opens in in lightbox
         * @param {?} galleryRef
         * @return {?}
         */
            function (galleryRef) {
                var _this = this;
                // Clone its items to the new gallery instance
                this._itemClick$ = this._galleryCmp.galleryRef.itemClick.subscribe(function (i) { return _this._lightbox.open(i, _this._galleryId); });
                this._itemChange$ = this._galleryCmp.galleryRef.itemsChanged.subscribe(function (state) { return galleryRef.load(state.items); });
            };
        /**
         * Detects images and adds a click event to each image to make it opens in the lightbox
         * @param {?} galleryRef
         * @return {?}
         */
        GallerizeDirective.prototype.detectorMode = /**
         * Detects images and adds a click event to each image to make it opens in the lightbox
         * @param {?} galleryRef
         * @return {?}
         */
            function (galleryRef) {
                var _this = this;
                this._detector$ = new rxjs.Subject();
                // Query image elements
                this._detector$.pipe(operators.debounceTime(300), operators.switchMap(function () {
                    /** *
                     * get all img elements from content
                      @type {?} */
                    var imageElements = _this._el.nativeElement.querySelectorAll(_this.selector);
                    if (imageElements && imageElements.length) {
                        /** @type {?} */
                        var images_1 = [];
                        return rxjs.from(imageElements).pipe(operators.map(function (el, i) {
                            // Add click event to the image
                            // Add click event to the image
                            _this._renderer.setStyle(el, 'cursor', 'pointer');
                            _this._renderer.setProperty(el, 'onclick', function () { return _this._lightbox.open(i, _this._galleryId); });
                            if (el instanceof HTMLImageElement) {
                                // If element is type of img use the src property
                                return {
                                    src: el.src,
                                    thumb: el.src
                                };
                            }
                            else {
                                /** @type {?} */
                                var elStyle = el.currentStyle || window.getComputedStyle(el, null);
                                /** @type {?} */
                                var background = elStyle.backgroundImage.slice(4, -1).replace(/"/g, '');
                                return {
                                    src: background,
                                    thumb: background
                                };
                            }
                        }), operators.tap(function (data) { return images_1.push(new core$1.ImageItem(data)); }), operators.finalize(function () { return galleryRef.load(images_1); }));
                    }
                    else {
                        return rxjs.EMPTY;
                    }
                })).subscribe();
                // Observe content changes
                this._observer$ = new MutationObserver(function () { return _this._detector$.next(); });
                this._observer$.observe(this._el.nativeElement, { childList: true, subtree: true });
            };
        GallerizeDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[gallerize]'
                    },] },
        ];
        /** @nocollapse */
        GallerizeDirective.ctorParameters = function () {
            return [
                { type: core.ElementRef },
                { type: core$1.Gallery },
                { type: lightbox.Lightbox },
                { type: core.Renderer2 },
                { type: Object, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
                { type: core$1.GalleryComponent, decorators: [{ type: core.Host }, { type: core.Self }, { type: core.Optional }] }
            ];
        };
        GallerizeDirective.propDecorators = {
            gallerize: [{ type: core.Input }],
            selector: [{ type: core.Input }]
        };
        return GallerizeDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var GallerizeModule = /** @class */ (function () {
        function GallerizeModule() {
        }
        GallerizeModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            core$1.GalleryModule,
                            lightbox.LightboxModule
                        ],
                        declarations: [GallerizeDirective],
                        exports: [GallerizeDirective]
                    },] },
        ];
        return GallerizeModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.GallerizeDirective = GallerizeDirective;
    exports.GallerizeModule = GallerizeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
