// @ngx-gallery/core/esm5/lib/components/templates/gallery-image.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, Output, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
var GalleryImageComponent = /** @class */ (function () {
    function GalleryImageComponent(_sanitizer) {
        this._sanitizer = _sanitizer;
        /**
         * Stream that emits when an error occurs
         */
        this.error = new EventEmitter();
    }
    Object.defineProperty(GalleryImageComponent.prototype, "imageLoaded", {
        get: /**
         * @return {?}
         */
        function () {
            return !!this.loadedImage;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    GalleryImageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.loadingIcon) {
            this.loadingSvg = this._sanitizer.bypassSecurityTrustHtml(this.loadingIcon);
        }
    };
    GalleryImageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'gallery-image',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: "\n    <ng-container [lazyImage]=\"src\" (loaded)=\"loadedImage = $event\" (error)=\"error.emit($event)\">\n      <div *ngIf=\"loadedImage; else loading\"\n           class=\"g-image-item\"\n           [style.backgroundImage]=\"loadedImage\"></div>\n    </ng-container>\n\n    <ng-template #loading>\n      <i *ngIf=\"loadingSvg; else thumbLoading\" class=\"g-loading\" [innerHTML]=\"loadingSvg\"></i>\n      <ng-template #thumbLoading>\n        <div class=\"g-thumb-loading\"></div>\n      </ng-template>\n    </ng-template>\n  "
                },] },
    ];
    /** @nocollapse */
    GalleryImageComponent.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
    GalleryImageComponent.propDecorators = {
        src: [{ type: Input }],
        loadingIcon: [{ type: Input }],
        error: [{ type: Output }],
        imageLoaded: [{ type: HostBinding, args: ['class.g-image-loaded',] }]
    };
    return GalleryImageComponent;
}());
export { GalleryImageComponent };
if (false) {
    /** @type {?} */
    GalleryImageComponent.prototype.loadedImage;
    /** @type {?} */
    GalleryImageComponent.prototype.loadingSvg;
    /** @type {?} */
    GalleryImageComponent.prototype.src;
    /** @type {?} */
    GalleryImageComponent.prototype.loadingIcon;
    /**
     * Stream that emits when an error occurs
     * @type {?}
     */
    GalleryImageComponent.prototype.error;
    /** @type {?} */
    GalleryImageComponent.prototype._sanitizer;
}
